sklearn/_fast_gradient_boosting/loss.pyx

```diff
@@ -0,0 +1,323 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+"""
+This module contains the loss classes.
+
+Specific losses are used for regression, binary classification or multiclass
+classification.
+"""
+from abc import ABC, abstractmethod
+
+cimport cython
+from cython.parallel import prange
+import numpy as np
+cimport numpy as np
+from scipy.special import expit
+try:
+    from scipy.special import logsumexp
+except ImportError:
+    from scipy.misc import logsumexp
+
+from libc.math cimport fabs, exp, log
+
+from .types import Y_DTYPE
+from .types cimport Y_DTYPE_C
+from .types import G_H_DTYPE
+from .types cimport G_H_DTYPE_C
+
+
+class BaseLoss(ABC):
+    """Base class for a loss."""
+
+    def init_gradients_and_hessians(self, n_samples, prediction_dim):
+        """Return initial gradients and hessians.
+
+        Unless hessians are constant, arrays are initialized with undefined
+        values.
+
+        Parameters
+        ----------
+        n_samples : int
+            The number of samples passed to `fit()`
+        prediction_dim : int
+            The dimension of a raw prediction, i.e. the number of trees
+            built at each iteration. Equals 1 for regression and binary
+            classification, or K where K is the number of classes for
+            multiclass classification.
+
+        Returns
+        -------
+        gradients : array-like, shape=(n_samples * prediction_dim)
+        hessians : array-like, shape=(n_samples * prediction_dim).
+            If hessians are constant (e.g. for ``LeastSquares`` loss, shape
+            is (1,) and the array is initialized to ``1``.
+        """
+        shape = n_samples * prediction_dim
+        gradients = np.empty(shape=shape, dtype=G_H_DTYPE)
```

Comment:
What was the motivation for using a single dimension instead of two dimensions for `gradients`? In `LeastSquares.update_gradients_and_hessians` and `BinaryCrossEntropy.update_gradients_and_hessians`, the `raw_predictions` needs to `reshape` to match the `gradients` shape. `_update_gradients_hessians_categorical_crossentropy` needs to jump around the array to update the gradient.